### ISSUE TITLE
Layer KeyBounds storage

### ISSUE THREAD

**Issue by pomadchin** [geotrellis/geotrellis#1291]
Currently `KeyBounds` stored for a certain layer, PR https://github.com/geotrellis/geotrellis/pull/1273 allows to store `KeyIndex` as a json. For `HilbertCurve` and for `RowMajor` indexes we have to store constructor arguments (in particular, `KeyBounds` is a part of it for these indexes). For layers indexed with this curves `KeyBounds` twice (for layer, in a layer `Metadata`, and for `KeyIndex`). 
May be we can use `KeyIndex.keyBounds` to store all _preallocated (James PR) / possible_ `KeyBounds` defined for a layer; and in `Metadata` to store _current (already ingested)_ `KeyBounds` for a layer? The only problem is a `ZCurve`, for it we should not store `KeyBounds` (that's the reason I am not sure in this approach).

```scala
trait KeyIndex[K] extends Serializable {
  val id: String = this.getClass.getName
  val keyBounds: KeyBounds[K] // key space for index
  def toIndex(key: K): Long 
  def indexRanges(keyRange: (K, K)): Seq[(Long, Long)]
}
```

**Comment by pomadchin** [comment]
@lossyrob @echeipesh @moradology @jamesmcclain any thoughts and suggestions?

**Comment by jamesmcclain** [comment]
I don't pretend to understand all of the implications of such a change, but I am generally in favor of reducing duplication.

**Comment by echeipesh** [comment]
There are a couple of changes going on related to this issue:

1. Soon `RasterMetaData` will contain `KeyBounds[K]` such that any layer read from a geotrellis catalog can be used with spatial partitioning. It's preferable to store all fields of an object together.

2. Now with upcoming ability to define an index for larger space than covered by the first layer write they two `KeyBounds` do not represent the same thing:

The `KeyBounds` stored with the Metadata represent the bounds of the data written and the data that can be read.
The `KeyBounds` stored with index represents indexable space.

So we have to have two instances of `KeyBounds` stored per layer

What you're saying is that there is no reason to store KeyBounds and KeyIndex in separate keys because they are always used together for the purpose of reading/writing a layer. Now that there is a more clear distinction between index and data KeyBounds this makes a lot more sense. You'd have to make a container object that holds KeyIndex and KeyBounds used for the index. 

Implied there is a question: Should `KeyIndex` interface contain `KeyBounds` ? It's required for construction of Hilbert and RowMajor index, but not required for Z. One can imagine that we can construct a Z index by off-setting keys by the `KeyBounds`, this would likely improve index compactness on average. I'm not sure that bounds belong in the index interface however, I think this is a point for discussion.

**Comment by jamesmcclain** [comment]
> What you're saying is that there is no reason to store KeyBounds and KeyIndex in separate keys because they are always used together for the purpose of reading/writing a layer. Now that there is a more clear distinction between index and data KeyBounds this makes a lot more sense. You'd have to make a container object that holds KeyIndex and KeyBounds used for the index.

From this paragraph I take the idea that it makes sense to keep the two `KeyBounds` together so that they are forced to have separate names which makes it clear that there are two separate (but related) concepts of `KeyBounds` in use.  That is useful, I think.

**Comment by pomadchin** [comment]
As we need `KeyBounds` for `KeyIndex` and in order not to spoil `KeyIndex` trait @echeipesh suggested the following trait:

```scala
trait BoundedKeyIndex[K] extends KeyIndex[K] {
   def bounds: KeyBounds[K]
}
```

**Comment by pomadchin** [comment]
https://github.com/pomadchin/geotrellis/commit/cb0d96ba4f0ec603eafd4b6c54322cc1f394bbf4 @echeipesh @jamesmcclain

**Comment by lossyrob** [comment]
Fixed by #1346
